src/common/runtime/helper/test_worker.ts

```diff
@@ -94,3 +94,53 @@ export class TestSharedWorker {
     rec.injectResult(workerResult);
   }
 }
+
+export class TestServiceWorker {
+  private readonly ctsOptions: CTSOptions;
+  private readonly resolvers = new Map<string, (result: LiveTestCaseResult) => void>();
+
+  constructor(ctsOptions?: CTSOptions) {
+    this.ctsOptions = { ...(ctsOptions || kDefaultCTSOptions), ...{ worker: 'service' } };
+  }
+
+  async run(
+    rec: TestCaseRecorder,
+    query: string,
+    expectations: TestQueryWithExpectation[] = []
+  ): Promise<void> {
+    const [suite, name] = query.split(':', 2);
+    const fileName = name.split(',').join('/');
+    const serviceWorkerPath = `/out/${suite}/webworker/${fileName}.worker.js`;
+
+    const registration = await navigator.serviceWorker.register(serviceWorkerPath, {
+      type: 'module',
+      scope: '/',
+    });
+    await registration.update();
+
+    navigator.serviceWorker.onmessage = ev => {
+      const query: string = ev.data.query;
+      const result: TransferredTestCaseResult = ev.data.result;
+      if (result.logs) {
+        for (const l of result.logs) {
+          Object.setPrototypeOf(l, LogMessageWithStack.prototype);
+        }
+      }
+      this.resolvers.get(query)!(result as LiveTestCaseResult);
+
+      // MAINTENANCE_TODO(kainino0x): update the Logger with this result (or don't have a logger and
+      // update the entire results JSON somehow at some point).
+    };
+
+    registration.active?.postMessage({
+      query,
+      expectations,
+      ctsOptions: this.ctsOptions,
+    });
+    const serviceWorkerResult = await new Promise<LiveTestCaseResult>(resolve => {
+      this.resolvers.set(query, resolve);
+      void registration.unregister();
+    });
+    rec.injectResult(serviceWorkerResult);
```

Comment:
I was not able to create a proper helper method for `run()` as we use either a Worker, MessagePort, or ServiceWorkerRegistration and we do different cleaning. Let me know if that's okay with you.

Reply:
It can't be the whole `run` method but everything from `postMessage` to `injectResult` is almost the same.
Here's a proposal which isn't actually simpler but I think it would be a bit more maintainable:
```ts
// also use this definition in wrapTestGroupForWorker and test_worker-worker.ts
interface WorkerTestRunRequest {
  query: string,
  expectations: TestQueryWithExpectation[] = [],
  ctsOptions: CTSOptions,
}
```
```ts
async makeRequestAndRecordResult(target: /* ? */, request: WorkerTestRunRequest) {
  target.postMessage({
    query,
    expectations,
    ctsOptions: this.ctsOptions,
  });
  const serviceWorkerResult = await new Promise<LiveTestCaseResult>(resolve => {
    this.resolvers.set(query, resolve);
  });
  rec.injectResult(serviceWorkerResult);
}
```
and for service worker:
```ts
await this.makeRequestAndRecordResult(/* ... */);
void registration.unregister();
```

Reply:
Note: `rec: TestCaseRecorder` is missing in `makeRequestAndRecordResult()`

Reply:
ah yes, that would need to be passed through too. Ends up being very verbose :(

Reply:
(I think it's probably still better though, for maintainability.)

Reply:
Done.